### ISSUE TITLE
Reassign § key to ~ in version 12

### ISSUE THREAD

**Issue by demalexx** [tekezo/Karabiner-Elements#1365]
Prior to version 12 I just chose ISO virtual keyboard and that's all I need to get "\~" (which is next to left shift) on proper place - instead of "§" (which is below Esc). Since version 12 there is no such setting. I tried to change country code, selected ISO in Keyboard Setup Assistant, but still I need to use "\~" key next to left shift to get my tilde.
And I don't have `System Preferences > Keyboard > Change Keyboard Type...` button.
Is it possible to have same behavior in version 12 as in older version?

**Comment by cotfas** [comment]
<img width="1011" alt="screenshot at apr 15 5-28-00 pm" src="https://user-images.githubusercontent.com/3340367/39072284-bfa40750-44f2-11e8-9f02-0e95d1c9b770.png">

I`m not sure, but maybe it can help you, I have the Karabinier 12, and I was able to switch ~ with § and left shift.

**Comment by dpwrussell** [comment]
I've just hit this myself. I'm not sure why it's happening, but on my ISO keyboard `§±` and `` `~ ``are reversed without any simple modifications in karabiner-elements.


When I push `§±`:
```
eventType:key_down        code:0x64       name:non_us_backslas misc:
eventType:key_up          code:0x64       name:non_us_backslas misc:
```
When I push `` `~``:
eventType:key_down        code:0x35       name:grave_accent_an misc:
eventType:key_up          code:0x35       name:grave_accent_an misc:
```

When Karabiner is not running, the keys output the correct symbols for an ISO (in my case UK) keyboard.

As others have said, it's not possible to `Change Keyboard Type` if either an external keyboard isn't attached or Karabine Elements is not running. If I do this when Karabiner Elements is running and let it autodetect by pressing the button to the right of the shift key, the mapping is still incorrect.

**I was able to circumvent this by tricking the autodetect when using `Change Keyboard Type`. I instead press the `§±` key when autodetecting and the detection results in a correctly mapped keyboard.**

Karabiner Elements version: 12.0.0

My intuition about high Karabiner Elements works is that it presents the mapped keyboard to the OS as a device (which is why the Keyboard Type button appears when Karabiner is launched which normally only appears for external keyboards) and that somehow these keys have got muddled in that simulated device.

Thanks. Karabiner is awesome!

**Comment by dpwrussell** [comment]
Actually, this isn't as easy to fix as I thought. I was using my laptop on its own, then plugged it back into the external keyboard, the keys had gotten switched back over and when I went to `Change Keyboard Type` and pressed a key on my external keyboard it said this:
![screen shot 2018-04-23 at 14 48 15](https://user-images.githubusercontent.com/1889043/39146770-b2df4e3e-4705-11e8-8509-6eb9c9a50234.png)

**Comment by e-gaulue** [comment]
+1

**Comment by luketsd** [comment]
+1, I thought it was maybe because I had previously plugged in a PC layout keyboard, and then gone back to a mac layout keyboard

**Comment by e-gaulue** [comment]
So, to sum it up. Since version 12:

1. some users have swapped keys when Karabiner is running (for me French keyboard: right key of left shift exchanged with above key of tabulation). 
2. on Karabiner installation, we had the "Change keyboard type" on Keyboard settings and the Keyboard Setup Assistant opened. We have set this (often to ISO in Europe)
3. then with time, can't say exactly when, the "Change keyboard type" button has disappeared.

Solution I'm testing is:
`defaults read /Library/Preferences/com.apple.HIToolbox.plist  AppleEnabledInputSources` then get the `KeyboardLayout ID` and put it in the Virtual Keyboard tab or Karabiner (was 0 before). Then Keyboard Setup Assistant opens again. Follow it. 

Until now, this works for me.

**Comment by demalexx** [comment]
@e-gaulue didn't work for me. So I created config to put tilda to correct location:

```json
{
    "title": "My rules",
    "rules": [
        {  
            "description": "Fix tilda location (from next from Shift to above Tab)",
            "manipulators": [
                {
                    "type": "basic",
                    "from": {
                        "key_code": "grave_accent_and_tilde",
                        "modifiers": {
                            "optional": [
                                "left_shift",
                                "right_shift"
                            ]
                        }
                    },
                    "to": [
                        {
                            "key_code": "non_us_backslash"
                        }
                    ]
                },
                {
                    "type": "basic",
                    "from": {
                        "key_code": "non_us_backslash",
                        "modifiers": {
                            "optional": [
                                "left_shift",
                                "right_shift"
                            ]
                        }
                    },
                    "to": [
                        {
                            "key_code": "grave_accent_and_tilde"
                        }
                    ]
                }
            ]
        }
    ]
}
```

**Comment by j0hnsmith** [comment]
A year or more ago I had § & ~ switched on a UK apple keyboard (bluetooth I think), the normal keyboard on my laptop was fine. This fixed it for me, well looking at it now maybe only half fixed it as I don't use the § key.

![image](https://www.dropbox.com/s/kdwali1ulmmc53w/2vnj01xm.png?raw=1)

**Comment by e-gaulue** [comment]
After a while, my trick as failed. But, here is a new one that leaves you reset your keyboard type to your need, all the time (with or without button on OSX keyboard settings).

First, I checked my keyboard type :
`defaults read /Library/Preferences/com.apple.keyboardtype "keyboardtype"`

Answer for me was : 
```
{
    "10203-5824-0" = 43;
}
```
You can have more than one: that's all the associations OS X has seen for your keyboard depending to its country code (the last digit `-0` is the same as the Country Code in the `Virtual keyboard` tab of Karabiner-Elements). 

So, as you can see my value for Country Code 0 is 43. In this case, the button to change keyboard type in OS X settings is hidden or has no effect when clicking on it.

In my case, I need 41. So I've changed the value this way: `sudo defaults write /Library/Preferences/com.apple.keyboardtype "keyboardtype" -dict "10203-5824-0" 41` but it works with anything other than 43. That's just to say something has changed.

As far as I understood, we have just modified the settings in the file, not in the computer memory. To make it works, we need to (don't ask me why) restart karabiner. Then you can click `change keyboard type` in the OSX keyboard settings.

In this case, the assistant opens again and leaves you choose your keyboard type. If you don't restart Karabiner, it does not. Looks like Karabiner keeps a lock on this setting.

BEWARE: Changing my keyboard type again from OS X settings got me back to 43 value, each time I did it. In this case, you just have to redo this trick again.

To sum it up:
1. Change value for your keyboard in /Library/Preferences/com.apple.keyboardtype to anything different than 43
2. Restart Karabiner
3. Change your keyboard type in the OSX settings

Regards,

**Comment by dpwrussell** [comment]
I don't think it is fixed yet.

**Comment by sveinugu** [comment]
Still not fixed, as far as I can see!

**Comment by rogual** [comment]
This bug still happens for me on macOS 10.14.6 with Karabiner-Elements 12.9.0.

**Comment by MTCoster** [comment]
Be very careful with the proposed fix above!

On my system:
```
➜ defaults read /Library/Preferences/com.apple.keyboardtype "keyboardtype"
{
    "10203-5824-0" = 43;
    "28673-4389-0" = 41;
    "45081-1133-0" = 41;
    "50475-1133-0" = 40;
    "9267-6551-0" = 41;
}
```

When using [the command above](https://github.com/pqrs-org/Karabiner-Elements/issues/1365#issuecomment-392009895), the remaining entries are overwritten:

```
➜ defaults write /Library/Preferences/com.apple.keyboardtype "keyboardtype" -dict "10203-5824-0" 41
➜ defaults read /Library/Preferences/com.apple.keyboardtype "keyboardtype"
{
    "10203-5824-0" = 41;
}
```

This caused my Mac to get stuck in a login loop, where the system crashed to the login screen every time Karabiner-Elements auto-loaded at login.

If anyone else hit this problem, restart while holding <kbd><kbd>⌘</kbd> <kbd>R</kbd></kbd> at boot to enter recovery mode. Delete (or rename) the `keyboardtype` plist file from the mounted Macintosh HD:

```
➜ rm -f /Volumes/Macintosh\ HD/Library/Preferences/com.apple.keyboardtype.plist
```

Then reboot. macOS will regenerate the missing file at login, after prompting with the Keyboard Setup Assistant:

```
➜ defaults read /Library/Preferences/com.apple.keyboardtype "keyboardtype"
{
    "10203-5824-0" = 41;
    "45081-1133-0" = 41;
    "9267-6551-0" = 41;
}
```

**Comment by richdel** [comment]
@MTCoster I had exactly the same issue and a similar output to yours running the `defaults read` cmd. In fact, I completely forgot Karabiner Elements was installed and have been scratching my head for days as to why the backtick key was swapped with +/- on a recent European/UK MBP retina device, after trying several times with keyboard setup with no success. Then I remembered I tried an external keyboard last week and needed some modifiers...

I've managed to fix this now, with Karabiner 14.2 (just ran an update today, but that alone didn't solve it). Running the `defaults read` cmd the below was output:

`> defaults read /Library/Preferences/com.apple.keyboardtype "keyboardtype"
{
    "10203-5824-0" = 41;
    "2-7247-0" = 41;
    "64097-9639-0" = 41;
    "8467-16700-0" = 43;
}`

I observed than when quitting KE the keyboard starts working as expected, and based on the above output I assumed the country code I was looking for was '41', so in KE > Preferences > Virtual Keyboard I simply entered that value there (rather than updating underlying OS X preferences manually):

<img width="734" alt="image" src="https://user-images.githubusercontent.com/138916/142253701-8bf60ce2-5031-46d1-8633-8461c2ec05b0.png">

After entering the value:

1. press the Enter/Return key on the keyboard
2. the keyboard setup assistant opens
3. follow the process as normal
4. close KE Preferences
5. hey presto everything works as expected

When running the `defaults read` cmd again there's now a new entry:

`{
    "10203-5824-0" = 41;
    **"10203-5824-41" = 43;**
    "2-7247-0" = 41;
    "64097-9639-0" = 41;
    "8467-16700-0" = 43;
}`

I don't know the mechanics of how this kbd prefs work so can't explain why, but it does :D Hope that helps!

**Comment by davepivonka** [comment]
@e-gaulue I have just tried your fix from 2018 - my computer crashed and my browser got completely reset - logged me out of all accounts, history got deleted etc... Should have checked newer possible solutions first haha

**Comment by davepivonka** [comment]
@MTCoster Oh well, I tried the solution before seeing your reply. Luckily, now it "only" crashed my computer whenever I started typing on the keyboard, so when I realised it after a few crashes, I had time to turn Karabiner Elements off with my trackpad. Thanks for letting us know the fix to the fix though haha

**Comment by dmayle** [comment]
I ran into this issue, and had some notes for if anyone else comes across it.

First, they keys in the plist, like 4369-4369-0 can be read in the System Information app by finding the keyboard under USB.  I have the Matias Ergo Pro, which is "Product ID: 0x1111" "Vendor ID: 0x1111". 0x1111 converted from hex to decimal is 4369.

Second, using defaults write, you want normally want to use `-dict-add`, which allows you to overwrite a single key/value in that dict.

Finally, in my case, changing the value to 41 for ISO on my keyboard caused WindowServer to crash every time I changed the value.  I decided, instead, to delete just the single entry for my keyboard with the value 43.

There is no `-dict-del`, so instead you have to use `-dict` and specify every single OTHER key value pair except for the one you want to delete.

For example:
```
sudo defaults write /Library/Preferences/com.apple.keyboardtype keyboardtype -dict 20548-12815-0 40 49288-1133-0 40 55-5426-0 42 73-9610-0 40
```
